README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# FeatureCorr

<!-- badges: start -->

[![R-CMD-check](https://github.com/DKundnani/FeatureCorr/actions/workflows/check-standard.yaml/badge.svg)](https://github.com/DKundnani/FeatureCorr/actions/workflows/check-standard.yaml)
[![Open in Code Ocean](https://codeocean.com/codeocean-assets/badge/open-in-code-ocean.svg)](https://codeocean.com/capsule/7549295/tree)
<!-- badges: end -->

An R package to study feature correlations aided with data
transformation for Next Generation sequencing and microarray data

## Installation

Get FeatureCorr from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("DKundnani/FeatureCorr")
```

## Example

Following is an example of function usage by utilizing sample data
provided in the package. It contains Gene-Tissue Expression dataset with
\~8000 genes(features) expression in 53 tissues, sampled from a larger
dataset containing \>42K genes.

Another dataset is from The Cancer Genome Atlas Pan Cancer dataset
randomly selected 2000 samples from \>10K samples, with information for
selected 40 features(genes)

``` r
library(FeatureCorr) #Loading the library
```

Data Transformation using FeatureCorr::data\_transform function

``` r
#Data Transformation
transdf<- data_transform(df=GTEX[-1],transformation='log2', featurelist=GTEX$Description, medianthres=1)
```

<img src="man/figures/README-preprocessing-1.png" width="100%" /> Data
Transformation using FeatureCorr::data\_transform function Prime Feature
Correlation using FeatureCorr::primefeature\_corr function

``` r

# Prime Feature(RNASEH2A gene) Correlation in Gene-Tissue EXpression Dataset
inputdf<-transdf[[1]] #First item in the list returned is the transformed Data
primefcorr <- primefeature_corr(df=inputdf,featurelist=rownames(inputdf) ,primefeature="RNASEH2A")
#> Calculating Correlation Coefficients
#> Calculating Distribution Statistics
#> Plotting Correlation Coefficient Distribution
```

<img src="man/figures/README-prime-feature-correlation-workflow-1.png" width="100%" />

Visualizing Single Pair Correlation using using
FeatureCorr::pair\_scatter function

``` r

#Single Pair Scatter on Original RNA-seq Dataset vs log2 transformed RNA-seq Dataset
pair_scatter(df=TCGA40,featurelist=rownames(TCGA40),feature1="RNASEH2A",feature2="PCNA", corrmeth='pearson')[[1]]
#> `geom_smooth()` using formula 'y ~ x'
#> `geom_smooth()` using formula 'y ~ x'
```

<img src="man/figures/README-checking-pairwise-scatter-1.png" width="100%" /><img src="man/figures/README-checking-pairwise-scatter-2.png" width="100%" />

``` r

pair_scatter(df=logTCGA40,featurelist=rownames(logTCGA40),feature1="RNASEH2A", feature2="PCNA", corrmeth='pearson')[[1]]
#> `geom_smooth()` using formula 'y ~ x'
#> `geom_smooth()` using formula 'y ~ x'
```

<img src="man/figures/README-checking-pairwise-scatter-3.png" width="100%" /><img src="man/figures/README-checking-pairwise-scatter-4.png" width="100%" />

Visualizing Multiple Pair-wise Correlation using
FeatureCorr::pairwise\_corr function

``` r

#Multiple pairwise correlation for 40 genes in ~10K samples in The Cancer Genome Atlas-Pan Cancer dataset
corr_pair<-pairwise_corr(df=logTCGA40,featurelist=rownames(logTCGA40), visorder="hclust")
```

<img src="man/figures/README-multiple-pairwise-correlation-1.png" width="100%" />

``` r

# Separating feature of interest for visualization 
# In This technique you can group features of differen types. For example if I have methylation and transcription data in the same dataframe for same set of feature identfiers
featgroup<-grepl( "RNASE",rownames(logTCGA40)) #optional, a set of features to separated
corr_pair<-pairwisecorr <- pairwise_corr(df=logTCGA40,featurelist=rownames(logTCGA40),featuregroup=featgroup)
```

<img src="man/figures/README-multiple-pairwise-correlation-2.png" width="100%" />


NEWS.md
# FeatureCorr 0.99.0

* `data_transform()` for Data Transformation and filtration

* `primefeature_corr()` Gets Correlation Coefficient and its distribution of a single prime feature with all the other features

* `pairwise_corr()` Gets Pair wise correlations for a given list of features

* `pair_scatter()` Scatter, dsitribution and correlation between two chosen features.

* Imports and Dependencies R (>= 4.0.0), GeneNet, ggExtra, MASS, ggpubr, corrplot, Hmisc, ggplot2, rlang, dplyr, KernSmooth, BiocManager


R/Misc.R
ztransfun<-function(col){
  (col-mean(as.numeric(unlist(col))))/sd(as.numeric(unlist(col)))
}
#' Data Transformation and filtration
#' Transformation of data using one of the four mentioned methods and filtering based on median value
#'
#'
#' @param df numerical dataframe with rows having series of values for a single feature
#' @param transformation type of transformation method ''Log2' or 'Z-score' or 'quantile' or 'NA' (no) transformaiton
#' @param featurelist listing of feature names or ids to measure correlation between
#' @param medianthres (optional)features have median below this numerical value are filtered or removed
#'
#' @usage data_transform(df,transformation,featurelist, medianthres)
#'
#' @examples
#' transdf<- data_transform(df=GTEX[-1],transformation='log2',
#'                          featurelist=GTEX$Description)
#' transdf<- data_transform(df=GTEX[-1],transformation='log2',
#'                          featurelist=GTEX$Description, medianthres=1)
#'
#' @export

data_transform <- function(df,transformation='log2',featurelist, medianthres='NA'){
  if (transformation == 'log2') {
    newdf=log2(df+1)
  } else if (transformation == 'Z-score') {
    newdf<-apply(df, 2, ztransfun)
  } else if (transformation == 'quantile') {
    if (!requireNamespace("BiocManager", quietly = TRUE))
      install.packages("BiocManager")
    BiocManager::install(c("preprocessCore"))
    newdf<-normalize.quantiles(as.matrix(df))
  } else if (transformation == 'NA') {
    newdf<-normalize.quantiles(as.matrix(df))
  } else {stop("There is no such transformation available")}

  colnames(newdf)<-colnames(df)
  rownames(newdf)<-featurelist

  OldMedianValue<-apply(df, 1, median, na.rm=F)
  NewMedianValue<-apply(newdf, 1, median, na.rm=F)
  Median=data.frame("MedianValues of Original DAtaframe" = OldMedianValue,
                    "MedianValues of Transformed DAtaframe" = NewMedianValue)

  if (medianthres!='NA'){
    newdf=newdf[NewMedianValue>=medianthres,]
    NewMedianValue<-NewMedianValue[NewMedianValue>=medianthres]
  }

  if (ncol(df)<15) {
    par(mfrow=c(2, 2), mar=c(15,5,2,2))
  } else {
    par(mfrow=c(1, 2), mar=c(5,5,2,2))
  }
  Om=hist(OldMedianValue, breaks = 200, xlab = "Median Value", col="chocolate",lty="blank", main ="Original Data")
  On=hist(NewMedianValue, breaks = 200, xlab = "Median Value", col="chocolate",lty="blank", main ="Transformed Data")

  if (ncol(df)<15) {
    OG.box<-boxplot(df, main ="Boxplots of Original Data", xaxt = "n")
    axis(side = 1,las = 2, labels = FALSE)
    text(x = 1:ncol(df),y = par("usr")[3] - 0.45,labels = colnames(df), srt = 90, adj=1.1, xpd = NA)

    if (medianthres!='NA') {
      new.box<-boxplot(newdf[ , -which(names(newdf) %in% c("feature"))], main ="Boxplots of Transformed Data", xaxt = "n")
      axis(side = 1,las = 2, labels = FALSE)
      text(x = 1:ncol(df),y = par("usr")[3] - 0.45,labels = colnames(df), srt = 90, adj=1.1, xpd = NA)
    } else {
      new.box<-boxplot(newdf, main ="Boxplots of Transformed Data", xaxt = "n")
      axis(side = 1,las = 2, labels = FALSE)
      text(x = 1:ncol(df),y = par("usr")[3] - 0.45,labels = colnames(df), srt = 90, adj=1.1, xpd = NA)
    }

  }

  dist.plot<- recordPlot()

  return(list(newdf,Median,dist.plot))

}







R/data.R
#' Gene-Tissue Expression (GTEx) Dataset v7
#'
#' A dataset containing the expression of >42000 genes in 54 human tissues
#'
#' @format A data frame with 8000 rows and 54 columns:
#'
#' @source \url{https://www.genome.gov/Funded-Programs-Projects/Genotype-Tissue-Expression-Project}
"GTEX"

#' The Cancer Genome Atlas, Pan Cancer Study Dataset for selected genes
#'
#' A dataset containing the expression of
#' 40 genes in 10071 human cancer samples(tissues or cells)
#'
#' @format A data frame with 40 rows and 2000 columns:
#'
#' @source \url{https://www.cbioportal.org/}
"TCGA40"

#' The Cancer Genome Atlas, Pan Cancer Study Dataset for selected genes(log transformed)
#'
#' Log 2 transformed dataset containing the expression of
#' 40 genes in 2000 human cancer samples(tissues or cells)
#'
#' @format A data frame with 40 rows and 10071 columns:
#'
"logTCGA40"


R/corrs.R
#Corr

#' Prime Feature Correlation
#' Get Correlation Coefficient and its distribution of a single prime feature with all the other features
#'
#' @import dplyr
#' @import rlang
#' @importFrom Hmisc rcorr
#' @import corrplot
#' @import ggpubr
#' @importFrom MASS kde2d
#' @import ggExtra
#' @import GeneNet
#'
#'
#' @param df A dataframe with echo row corresponding to a feature
#' @param featurelist listing of feature names or ids to measure correlation between
#' @param primefeature one feature of interest to measure correlation against all other features
#' @param corrmeth Correlation method used. 'pearson' or 'spearman' (default='pearson')
#' @param quantiles list of corresponding probabilities for quantiles to be obtained
#' (default = c(0,0.01,.01,.05,0.10,.25,.50,.75,.90,.95,.99,.999,1) )
#'
#' @examples
#' transdf<- data_transform(df=GTEX[-1],transformation='log2',
#'                          featurelist=GTEX$Description, medianthres=1)
#' inputdf<-transdf[[1]] #First item in the list returned is the transformed Data
#' primefcorr <- primefeature_corr(df=inputdf,featurelist=rownames(inputdf),primefeature="RNASEH2A");
#'
#' @usage primefeature_corr(df,featurelist,primefeature,corrmeth,quantiles)
#'
#' @export
primefeature_corr <- function(df,featurelist,primefeature, corrmeth='pearson', quantiles=c(0,0.01,.01,.05,0.10,.25,.50,.75,.90,.95,.99,.999,1)){
  if (length(featurelist)!=nrow(df) && length(primefeature)>1) { stop("Error: Make sure the length of feature list is same as the number of rows in the dataframe and there is only one prime feature.n", call. = FALSE) }
  c=ncol(df)
  r=nrow(df)
  corr<-rep(NA, r)
  pval<-rep(NA, r)


  x=match(primefeature, featurelist)
  if (is.na(x)) {  stop("Primefeature not found in feature list.n", call. = FALSE)}

  writeLines("Calculating Correlation Coefficients")
  for (i in 1:r){

    p<-cor.test(as.numeric(df[i,]), as.numeric(df[x,]), method=corrmeth)
    corr[i]<-p$estimate
    pval<-p$p.value
  }

  df$Features<-featurelist
  df$Correlation_Coefficient<-corr
  df$Correlation_pvalue<-pval

  correlations<-df[,(c+1):ncol(df)]

  writeLines("Calculating Distribution Statistics")

  quantiles<-quantile(df$Correlation_Coefficient, quantiles)


  writeLines("Plotting Correlation Coefficient Distribution")

  par(mfrow=c(1, 1), mar=c(7,5,1,1))
  s=hist(df$Correlation_Coefficient, breaks = 200, xlab = "Correlation Coefficient", col="steelblue3", main ="Distribution of Correlation Coefficient")

  freq.plot<- recordPlot()

  finaloutput<-list(df,correlations, quantiles,s,freq.plot)

  names(finaloutput) <- c("DatawithCorr", "Correlations","Quantiles","Corr.freq",  "Frequency_Dist")
  return(finaloutput)
}


#' Pairwise Correlation
#'
#' Get Pair wise correlations for a given list of features
#' @param df numerical dataframe with rows having series of values for a single feature
#' @param featurelist list of features in the sequence as it occurs in the Dataframe
#' @param featuregroup (optional)list of TRUE/FALSE values for features to be observed vs rest of the features(ordering/clustering is disabled)
#' @param visorder  The ordering method of the correlation matrix.'original' for original order ,'AOE' for the angular order of the eigenvectors, 'FPC' for the first principal component order, 'hclust'(default) for the hierarchical clustering order, 'alphabet' for alphabetical order.
#' @param corrmeth Correlation method used. 'pearson' or 'spearman' (default='pearson')
#'
#' @examples
#' corr_pair<-pairwise_corr(df=logTCGA40,featurelist=rownames(logTCGA40), visorder="hclust")
#' featgroup<-grepl( "RNASE",rownames(logTCGA40)) #optional, a set of TRUE/FALSE of length featurelist
#' corr_pair<-pairwisecorr <- pairwise_corr(df=logTCGA40,featurelist=rownames(logTCGA40),
#'                                          featuregroup=featgroup)
#'
#' @usage pairwise_corr(df,featurelist, featuregroup,visorder,corrmeth)
#'
#' @export
pairwise_corr <- function(df,featurelist, featuregroup='NA', visorder="hclust", corrmeth='pearson'){
  if (length(featurelist)!=nrow(df)) { stop("Error: Make sure the length of feature list is same as the number of rows in the dataframe", call. = FALSE) }
  rownames(df)<-featurelist
  corr_pair<-rcorr(t(as.matrix(df)), type=corrmeth)
  par(mfrow=c(1, 1), mar=c(1,4,4,8))

  if (length(featuregroup) == 1) {
    corrplot(corr_pair$r,type = "full", order =  "hclust",addrect = 2,method = "square",tl.col = "black",tl.srt = 60, cl.cex = 1.3, pch.cex = 0.5, tl.cex = 0.7)
    out=list(corr_pair$r,corr_pair$P)

  } else {
    corrplot(corr_pair$r[featuregroup,!featuregroup],method = "square",tl.col = "black",tl.srt = 90, cl.cex = 1.3, pch.cex = 0.5, is.corr = FALSE, cl.pos = 'b', cl.ratio = 0.6)
    out=list(corr_pair$r,corr_pair$P)

  }
  return(out)

}

#' Pair Scatter
#' Scatter between two chosen features.
#'
#' @import ggpubr
#' @import ggplot2
#'
#'
#' @param df numerical dataframe with rows having series of values for a single feature
#' @param featurelist list of features in the sequence as it occurs in the Dataframe
#' @param feature1,feature2 Features to be compared
#' @param corrmeth Correlation method used. 'pearson' or 'spearman' (default='pearson')
#'
#'
#' @usage pair_scatter(df,featurelist, feature1, feature2, corrmeth)
#'
#' @examples
#' pscatter <- pair_scatter(df=TCGA40,featurelist=rownames(TCGA40),
#'                          feature1="RNASEH2A",feature2="PCNA", corrmeth='pearson')
#' pscatter <- pair_scatter(df=logTCGA40,featurelist=rownames(logTCGA40),
#'                          feature1="RNASEH2A", feature2="PCNA", corrmeth='pearson')
#'
#' @export
pair_scatter <- function(df,featurelist, feature1, feature2,corrmeth='pearson'){
  if (length(featurelist)!=nrow(df)) { stop("Error: Make sure the length of feature list is same as the number of rows in the dataframe", call. = FALSE) }

  data=as.data.frame(t(df))
  colnames(data)<-featurelist
  #scatter plot
  #par(mfrow = c(1, 1))
  o <- ggscatter(data, x = feature1 , y = feature2, color = "#2c7fb8", margin.params = list(fill = "lightgray"), add = "reg.line", add.params = list(color='red', fille='lightgray'),conf.int = TRUE, cor.coef = TRUE, cor.coeff.args = list(method=corrmeth,label.x.npc = 0,label.y.npc = 1, size=10)) +theme_bw()
  p <-ggMarginal(o, size = 2, type = "histogram", col = "blue", fill = "#2c7fb8")

  plot.new()
  kern <- kde2d(as.numeric(unlist(data[feature1])),as.numeric(unlist(data[feature2])))
  smoothScatter(as.numeric(unlist(data[feature1])),as.numeric(unlist(data[feature2])), xlab=feature1, ylab =feature2, nrpoints = 0)
  contour(kern, drawlabels = FALSE, nlevels = 6,
          col = rev(heat.colors(6)), add = TRUE, lwd = 3)
  q <- recordPlot()

  ## clean up device
  plot.new()

  return(list(p,q))
}



